File: php/src/MlKem/Kpke.php
<?php

declare(strict_types=1);

namespace PQC\MlKem;

/**
 * K-PKE: The internal public-key encryption scheme used in ML-KEM.
 * FIPS 203 Algorithms 12-14.
 */
final class Kpke
{
    /**
     * K-PKE.KeyGen: Generate encryption key pair.
     * FIPS 203 Algorithm 12.
     *
     * @param string $d 32-byte seed
     * @param int $level 512, 768, or 1024
     * @return array{ek: string, dk: string}
     */
    public static function keyGen(string $d, int $level): array
    {
        $params = Params::get($level);
        $k = $params['k'];
        $eta1 = $params['eta1'];

        // (rho, sigma) = G(d || k)
        $gs = HashFuncs::G($d . chr($k));
        $rho = substr($gs, 0, 32);
        $sigma = substr($gs, 32, 32);

        // Generate matrix A (in NTT domain) from rho
        $aHat = [];
        for ($i = 0; $i < $k; $i++) {
            for ($j = 0; $j < $k; $j++) {
                $aHat[$i][$j] = Sampling::sampleNtt($rho, $i, $j);
            }
        }

        // Sample secret vector s
        $sVec = Sampling::sampleNoiseVec($sigma, $eta1, $k, 0);

        // Sample error vector e
        $eVec = Sampling::sampleNoiseVec($sigma, $eta1, $k, $k);

        // NTT(s), NTT(e)
        $sHat = [];
        $eHat = [];
        for ($i = 0; $i < $k; $i++) {
            $sHat[$i] = Ntt::ntt($sVec[$i]);
            $eHat[$i] = Ntt::ntt($eVec[$i]);
        }

        // t_hat = A_hat * s_hat + e_hat
        $tHat = [];
        for ($i = 0; $i < $k; $i++) {
            $tHat[$i] = $eHat[$i];
            for ($j = 0; $j < $k; $j++) {
                $prod = Ntt::mulNtt($aHat[$i][$j], $sHat[$j]);
                $tHat[$i] = Ntt::polyAdd($tHat[$i], $prod);
            }
        }

        // Encode ek = ByteEncode_12(t_hat) || rho
        $ek = '';
        for ($i = 0; $i < $k; $i++) {
            $ek .= Encode::byteEncode($tHat[$i], 12);
        }
        $ek .= $rho;

        // Encode dk = ByteEncode_12(s_hat)
        $dk = '';
        for ($i = 0; $i < $k; $i++) {
            $dk .= Encode::byteEncode($sHat[$i], 12);
        }

        return ['ek' => $ek, 'dk' => $dk];
    }

    /**
     * K-PKE.Encrypt: Encrypt a 32-byte message.
     * FIPS 203 Algorithm 13.
     *
     * @param string $ek Encryption key
     * @param string $m 32-byte message
     * @param string $r 32-byte randomness
     * @param int $level 512, 768, or 1024
     * @return string Ciphertext
     */
    public static function encrypt(string $ek, string $m, string $r, int $level): string
    {
        $params = Params::get($level);
        $k = $params['k'];
        $eta1 = $params['eta1'];
        $eta2 = $params['eta2'];
        $du = $params['du'];
        $dv = $params['dv'];

        // Decode ek
        $tHat = [];
        for ($i = 0; $i < $k; $i++) {
            $tHat[$i] = Encode::byteDecode(substr($ek, $i * 384, 384), 12);
        }
        $rho = substr($ek, $k * 384, 32);

        // Generate matrix A^T from rho
        $aHat = [];
        for ($i = 0; $i < $k; $i++) {
            for ($j = 0; $j < $k; $j++) {
                $aHat[$i][$j] = Sampling::sampleNtt($rho, $i, $j);
            }
        }

        // Sample vectors r, e1 and scalar e2
        $rVec = Sampling::sampleNoiseVec($r, $eta1, $k, 0);
        $e1Vec = Sampling::sampleNoiseVec($r, $eta2, $k, $k);
        $e2 = Sampling::sampleNoisePoly($r, $eta2, 2 * $k);

        // NTT(r)
        $rHat = [];
        for ($i = 0; $i < $k; $i++) {
            $rHat[$i] = Ntt::ntt($rVec[$i]);
        }

        // u = NTT^{-1}(A^T * r_hat) + e1
        $uVec = [];
        for ($i = 0; $i < $k; $i++) {
            $acc = array_fill(0, 256, 0);
            for ($j = 0; $j < $k; $j++) {
                $prod = Ntt::mulNtt($aHat[$j][$i], $rHat[$j]);
                $acc = Ntt::polyAdd($acc, $prod);
            }
            $uVec[$i] = Ntt::polyAdd(Ntt::invNtt($acc), $e1Vec[$i]);
        }

        // v = NTT^{-1}(t_hat^T * r_hat) + e2 + Decompress_1(m)
        $acc = array_fill(0, 256, 0);
        for ($i = 0; $i < $k; $i++) {
            $prod = Ntt::mulNtt($tHat[$i], $rHat[$i]);
            $acc = Ntt::polyAdd($acc, $prod);
        }
        $v = Ntt::invNtt($acc);
        $v = Ntt::polyAdd($v, $e2);

        // Decode message to polynomial
        $mPoly = Encode::byteDecode($m, 1);
        $mDecomp = Compress::decompressPoly($mPoly, 1);
        $v = Ntt::polyAdd($v, $mDecomp);

        // Compress and encode
        $c1 = '';
        for ($i = 0; $i < $k; $i++) {
            $c1 .= Compress::compressAndEncode($uVec[$i], $du);
        }
        $c2 = Compress::compressAndEncode($v, $dv);

        return $c1 . $c2;
    }

    /**
     * K-PKE.Decrypt: Decrypt a ciphertext.
     * FIPS 203 Algorithm 14.
     *
     * @param string $dk Decryption key
     * @param string $ct Ciphertext
     * @param int $level 512, 768, or 1024
     * @return string 32-byte decrypted message
     */
    public static function decrypt(string $dk, string $ct, int $level): string
    {
        $params = Params::get($level);
        $k = $params['k'];
        $du = $params['du'];
        $dv = $params['dv'];
        $sizes = Params::sizes($level);

        // Decode ciphertext
        $uVec = [];
        $offset = 0;
        for ($i = 0; $i < $k; $i++) {
            $uVec[$i] = Compress::decodeAndDecompress(
                substr($ct, $offset, $sizes['polyCompressedDu']),
                $du
            );
            $offset += $sizes['polyCompressedDu'];
        }
        $v = Compress::decodeAndDecompress(
            substr($ct, $offset, $sizes['polyCompressedDv']),
            $dv
        );

        // Decode secret key
        $sHat = [];
        for ($i = 0; $i < $k; $i++) {
            $sHat[$i] = Encode::byteDecode(substr($dk, $i * 384, 384), 12);
        }

        // m = v - NTT^{-1}(s_hat^T * NTT(u))
        $acc = array_fill(0, 256, 0);
        for ($i = 0; $i < $k; $i++) {
            $uHat = Ntt::ntt($uVec[$i]);
            $prod = Ntt::mulNtt($sHat[$i], $uHat);
            $acc = Ntt::polyAdd($acc, $prod);
        }
        $w = Ntt::invNtt($acc);
        $mPoly = Ntt::polySub($v, $w);

        // Compress to 1 bit and encode
        $mCompressed = Compress::compressPoly($mPoly, 1);
        return Encode::byteEncode($mCompressed, 1);
    }
}

File: go/mlkem/kpke.go
// Package mlkem implements ML-KEM (FIPS 203) key encapsulation mechanism.
package mlkem

import (
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/compress"
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/encode"
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/field"
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/hash"
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/ntt"
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/params"
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/sampling"
)

// KPKEKeyGen implements K-PKE.KeyGen (FIPS 203 Algorithm 13).
// It generates an encryption key and decryption key from a 32-byte seed d.
func KPKEKeyGen(p params.ParameterSet, d [32]byte) (ekPKE []byte, dkPKE []byte) {
	k := p.K

	// Line 1: (rho, sigma) = G(d)
	rho, sigma := hash.G(d[:])

	// Line 2-6: Generate matrix A_hat in NTT domain.
	// A_hat[i][j] = SampleNTT(rho || j || i)
	aHat := make([][]([256]field.Element), k)
	for i := 0; i < k; i++ {
		aHat[i] = make([][256]field.Element, k)
		for j := 0; j < k; j++ {
			var seed [34]byte
			copy(seed[:32], rho[:])
			seed[32] = byte(j)
			seed[33] = byte(i)
			aHat[i][j] = sampling.SampleNTT(seed)
		}
	}

	// Line 7-10: Sample secret vector s and compute s_hat = NTT(s).
	var n byte
	sHat := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		sHat[i] = sampling.SamplePolyCBD(p.ETA1, sampling.PRF(sigma, n, 64*p.ETA1))
		n++
		ntt.NTT(&sHat[i])
	}

	// Line 11-14: Sample error vector e and compute e_hat = NTT(e).
	eHat := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		eHat[i] = sampling.SamplePolyCBD(p.ETA1, sampling.PRF(sigma, n, 64*p.ETA1))
		n++
		ntt.NTT(&eHat[i])
	}

	// Line 15-16: t_hat = A_hat * s_hat + e_hat
	tHat := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		// t_hat[i] = sum_j(A_hat[i][j] * s_hat[j]) + e_hat[i]
		for j := 0; j < k; j++ {
			product := ntt.MultiplyNTTs(&aHat[i][j], &sHat[j])
			for c := 0; c < 256; c++ {
				tHat[i][c] = tHat[i][c].Add(product[c])
			}
		}
		for c := 0; c < 256; c++ {
			tHat[i][c] = tHat[i][c].Add(eHat[i][c])
		}
	}

	// Line 17: ekPKE = ByteEncode(12, t_hat[0]) || ... || ByteEncode(12, t_hat[k-1]) || rho
	ekPKE = make([]byte, 0, 384*k+32)
	for i := 0; i < k; i++ {
		ekPKE = append(ekPKE, encode.ByteEncode(12, &tHat[i])...)
	}
	ekPKE = append(ekPKE, rho[:]...)

	// Line 18: dkPKE = ByteEncode(12, s_hat[0]) || ... || ByteEncode(12, s_hat[k-1])
	dkPKE = make([]byte, 0, 384*k)
	for i := 0; i < k; i++ {
		dkPKE = append(dkPKE, encode.ByteEncode(12, &sHat[i])...)
	}

	return ekPKE, dkPKE
}

// KPKEEncrypt implements K-PKE.Encrypt (FIPS 203 Algorithm 14).
// It encrypts a 32-byte message m under encryption key ek using randomness r.
func KPKEEncrypt(p params.ParameterSet, ek []byte, m [32]byte, r [32]byte) []byte {
	k := p.K

	// Line 1-3: Decode t_hat from ek and extract rho.
	tHat := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		tHat[i] = encode.ByteDecode(12, ek[384*i:384*(i+1)])
	}
	var rho [32]byte
	copy(rho[:], ek[384*k:])

	// Line 4-8: Reconstruct A_hat from rho.
	aHat := make([][]([256]field.Element), k)
	for i := 0; i < k; i++ {
		aHat[i] = make([][256]field.Element, k)
		for j := 0; j < k; j++ {
			var seed [34]byte
			copy(seed[:32], rho[:])
			seed[32] = byte(j)
			seed[33] = byte(i)
			aHat[i][j] = sampling.SampleNTT(seed)
		}
	}

	// Line 9-12: Sample y vector and compute y_hat = NTT(y).
	var n byte
	yHat := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		yHat[i] = sampling.SamplePolyCBD(p.ETA1, sampling.PRF(r, n, 64*p.ETA1))
		n++
		ntt.NTT(&yHat[i])
	}

	// Line 13-16: Sample e1 (error vector, not NTT'd).
	e1 := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		e1[i] = sampling.SamplePolyCBD(p.ETA2, sampling.PRF(r, n, 64*p.ETA2))
		n++
	}

	// Line 17: Sample e2 (single error polynomial, not NTT'd).
	e2 := sampling.SamplePolyCBD(p.ETA2, sampling.PRF(r, n, 64*p.ETA2))
	// n++ not needed after this

	// Line 18-19: u = NTT^{-1}(A^T * y_hat) + e1
	// A^T[i][j] = A[j][i], so (A^T * y_hat)[i] = sum_j(A_hat[j][i] * y_hat[j])
	u := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		var uHat [256]field.Element
		for j := 0; j < k; j++ {
			product := ntt.MultiplyNTTs(&aHat[j][i], &yHat[j])
			for c := 0; c < 256; c++ {
				uHat[c] = uHat[c].Add(product[c])
			}
		}
		ntt.NTTInverse(&uHat)
		for c := 0; c < 256; c++ {
			u[i][c] = uHat[c].Add(e1[i][c])
		}
	}

	// Line 20: mu = Decompress(1, ByteDecode(1, m))
	mDecoded := encode.ByteDecode(1, m[:])
	var mu [256]field.Element
	for i := 0; i < 256; i++ {
		mu[i] = compress.Decompress(1, mDecoded[i].Value())
	}

	// Line 21: v = NTT^{-1}(t_hat^T * y_hat) + e2 + mu
	var vHat [256]field.Element
	for j := 0; j < k; j++ {
		product := ntt.MultiplyNTTs(&tHat[j], &yHat[j])
		for c := 0; c < 256; c++ {
			vHat[c] = vHat[c].Add(product[c])
		}
	}
	ntt.NTTInverse(&vHat)
	var v [256]field.Element
	for c := 0; c < 256; c++ {
		v[c] = vHat[c].Add(e2[c]).Add(mu[c])
	}

	// Line 22-24: Compress and encode u and v.
	ct := make([]byte, 0, p.CTSize)
	for i := 0; i < k; i++ {
		var compressed [256]field.Element
		for c := 0; c < 256; c++ {
			compressed[c] = field.Element(compress.Compress(p.DU, u[i][c]))
		}
		ct = append(ct, encode.ByteEncode(p.DU, &compressed)...)
	}

	var compressedV [256]field.Element
	for c := 0; c < 256; c++ {
		compressedV[c] = field.Element(compress.Compress(p.DV, v[c]))
	}
	ct = append(ct, encode.ByteEncode(p.DV, &compressedV)...)

	return ct
}

// KPKEDecrypt implements K-PKE.Decrypt (FIPS 203 Algorithm 15).
// It decrypts a ciphertext ct using decryption key dk.
func KPKEDecrypt(p params.ParameterSet, dk []byte, ct []byte) [32]byte {
	k := p.K

	// Line 1-4: Decompress u from ciphertext.
	u := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		uCompressed := encode.ByteDecode(p.DU, ct[32*p.DU*i:32*p.DU*(i+1)])
		for c := 0; c < 256; c++ {
			u[i][c] = compress.Decompress(p.DU, uCompressed[c].Value())
		}
	}

	// Line 5: Decompress v from ciphertext.
	vOffset := 32 * p.DU * k
	vCompressed := encode.ByteDecode(p.DV, ct[vOffset:vOffset+32*p.DV])
	var v [256]field.Element
	for c := 0; c < 256; c++ {
		v[c] = compress.Decompress(p.DV, vCompressed[c].Value())
	}

	// Line 6-7: Decode s_hat from dk.
	sHat := make([][256]field.Element, k)
	for i := 0; i < k; i++ {
		sHat[i] = encode.ByteDecode(12, dk[384*i:384*(i+1)])
	}

	// Line 8: w = v - NTT^{-1}(s_hat^T * NTT(u))
	// Compute s_hat^T * NTT(u) = sum_i(s_hat[i] * NTT(u[i]))
	var inner [256]field.Element
	for i := 0; i < k; i++ {
		uNTT := u[i]
		ntt.NTT(&uNTT)
		product := ntt.MultiplyNTTs(&sHat[i], &uNTT)
		for c := 0; c < 256; c++ {
			inner[c] = inner[c].Add(product[c])
		}
	}
	ntt.NTTInverse(&inner)

	var w [256]field.Element
	for c := 0; c < 256; c++ {
		w[c] = v[c].Sub(inner[c])
	}

	// Line 9: m = ByteEncode(1, Compress(1, w))
	var compressed [256]field.Element
	for c := 0; c < 256; c++ {
		compressed[c] = field.Element(compress.Compress(1, w[c]))
	}
	encoded := encode.ByteEncode(1, &compressed)

	var m [32]byte
	copy(m[:], encoded)
	return m
}
